### PR TITLE
Support multihost SPMD execution

### PR DESCRIPTION
**Pull request by jonb377**
The only main change to support multihost execution is to restrict the generated shards in ShardTensor to those which belong to addressable devices.

### CODE REVIEW

File: torch_xla/experimental/xla_sharding.py [hunk 1]
```diff
@@ -88,7 +89,7 @@ def mark_sharding(t: Union[torch.Tensor, XLAShardedTensor], mesh: Mesh,
     Examples
     —------------------------------
     mesh_shape = (4, 2)
-    num_devices = len(xm.get_xla_supported_devices())
+    num_devices = pjrt.global_device_count()
```
Comment (yeounoh) [inline torch_xla/experimental/xla_sharding.py:92]: Great :)

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 2]
```diff
@@ -175,26 +175,41 @@ xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
   return module.get()->ToProto();
 }
 
+// Builds a map from the device's global ordinal to its index in the `devices`
+// array.
+static std::unordered_map<int, int> build_index_map(
+    const std::vector<std::string>& devices) {
+  std::unordered_map<int, int> device_index;
+  for (int i = 0; i < devices.size(); ++i) {
+    int global_ordinal = ParseDeviceString(devices[i]).ordinal();
```
Comment (yeounoh) [inline torch_xla/csrc/xla_sharding_util.cpp:184]: The first global device gets the local index 0, so the order of the input devices list is important. Is this a correct understanding? Can we add some comments on this?

**Review by yeounoh — CHANGES_REQUESTED** [review]
LGTM, a minor comment.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 2 above)
Comment (jonb377) [reply torch_xla/csrc/xla_sharding_util.cpp:184]: The first device in the list gets local index 0, but the order of the global ordinals within `devices` doesn't matter. I'll add some more documentation around this.

File: torch_xla/csrc/tensor_util.cpp [hunk 3]
```diff
@@ -931,30 +931,24 @@ std::vector<torch::lazy::BackendDataPtr> CreateTensorsData(
     std::vector<xla::ComputationClient::DataPtr> new_handles;          // out
     if (shardings[i] != nullptr) {
       xla::OpSharding sharding = shardings[i]->sharding;
-      // TODO(yeounoh) PJRT runs a process per host for SPMD and without cross
-      // host communications. This means that we may need to manually shard
-      // across global devices for multi-host training.
       std::vector<std::string> local_devices =
```
Comment (yeounoh) [inline torch_xla/csrc/tensor_util.cpp:934]: Does `GetLocalDevices()` return local devices with global ordinals? If so, let's leave a comment.

**Review by yeounoh — APPROVED** [review]
LGTM, have 2 comments --nit.

**Pull request merged**